### PR TITLE
Load interface scripts as CommonJS modules, refactor Firefox Adapter

### PR DESCRIPTION
**Pull request by hitesh96db**
The changes include --
* Load the interface scripts as CommonJS modules so that they can be used in the jetpack background scripts.
* Added more functionality to the Firefox Adapter.
* Options interface bug fix.
* Changed the check for `Array` data-type in `options.js`.

### CODE REVIEW

File: shared/javascripts/options.js [hunk 1]
```diff
@@ -364,7 +377,8 @@ if (Privly === undefined) {
     if (typeof glyph !== 'object' || glyph === null) {
       throw new Error('invalid argument');
     }
-    if (typeof glyph.color !== 'string' || typeof glyph.cells !== 'object' || glyph.cells === null || glyph.cells.constructor !== Array) {
+    if (typeof glyph.color !== 'string' || typeof glyph.cells !== 'object' || 
+        glyph.cells === null || Object.prototype.toString.call(glyph.cells) !== "[object Array]") {
```
Comment (hitesh96db) [inline shared/javascripts/options.js]: `glyph.cells.constructor !== Array` will not work for CommonJS modules because the `Array` global object isn't the same one in two different modules/env.

**Comment by hitesh96db** [comment]
The changes in the extension can be found in this [commit](https://github.com/hitesh96db/privly-jetpack/commit/153a98459b7097b39e05d8ae6183300e919c146a) or in the [inbetween](https://github.com/hitesh96db/privly-jetpack/tree/inbetween)(probably not a nice/informative name) branch.

File: shared/javascripts/options.js [hunk 2]
```diff
@@ -54,6 +54,7 @@ var Privly;
 if (Privly === undefined) {
   Privly = {};
 }
+var ls;
```
Comment (smcgregor) [inline shared/javascripts/options.js]: Why is `ls` here?

File: shared/javascripts/options.js (hunk 1 above)
Comment (smcgregor) [inline shared/javascripts/options.js]: Notes like these are usually better left as a comment in the source code.

File: shared/javascripts/options.js (hunk 2 above)
Comment (hitesh96db) [inline shared/javascripts/options.js]: right, not needed.

File: shared/javascripts/options.js (hunk 1 above)
Comment (hitesh96db) [inline shared/javascripts/options.js]: Done. I've rebased the commits so the changes can't be seen as a separate commit.

**Comment by hitesh96db** [comment]
@smcgregor Hey, I've added a new commit to this PR. Is this Ok to merge ?

**Comment by smcgregor** [comment]
I reviewed the new commits, with one comment. Please respond to the comment and I'll run the unit tests on CI now.

**Comment by smcgregor** [comment]
There are some [syntax errors](https://travis-ci.org/smcgregor/privly-applications/builds/73712805).

**Comment by hitesh96db** [comment]
All the syntax errors are nearly the same thing, but I can't understand what's wrong.

**Comment by smcgregor** [comment]
I am re-running the build just in case it is a non-deterministic error.

**Comment by smcgregor** [comment]
Same failures, I am [running an old build](https://travis-ci.org/smcgregor/privly-applications/builds/67697755) to see if it is still passing.

**Comment by smcgregor** [comment]
The old build works. I think the issue may be the `const` declaration. Specifically, I don't think you can assign into a collection like this across platforms and it is messing with the parser: `const { Cc, Ci }`.

**Comment by hitesh96db** [comment]
Yeah, probably, I can change that..

**Comment by hitesh96db** [comment]
@smcgregor I've added the changes.

**Comment by smcgregor** [comment]
Build is running [here](https://travis-ci.org/smcgregor/privly-applications/builds/73743849). I'll work on merging if it passes.

**Comment by smcgregor** [comment]
It looks like you need to use a different name (see the build).

**Comment by hitesh96db** [comment]
@smcgregor I've pushed the changes.

**Comment by smcgregor** [comment]
[Running the build](https://travis-ci.org/smcgregor/privly-applications/builds/73922600). I had to push a different branch because the history of this PR changed.

File: shared/javascripts/context_messenger.js [hunk 3]
```diff
@@ -316,43 +321,66 @@ if (Privly === undefined) {
...
-    worker.port.on("privly", callback);
+    
+    // Add worker to array of workers
+    pushWorker(worker);
+    // Listen for messages.
+    worker.port.on("PRIVLY_MESSAGE", callback);
+    // the workers array should contain only active workers.
+    // remove the workers, that are inactive or destroyed, from the workers array.
+    worker.on("pageshow", function() { pushWorker(this); });
+    worker.on("pagehide", function() { popWorker(this); });
```
Comment (smcgregor) [inline shared/javascripts/context_messenger.js]: Do you need to remove the workers on hide and detach? This breaks deactivating injection, no?

Comment (hitesh96db) [inline shared/javascripts/context_messenger.js]: Nope, it'll work. The background script doesn't message the content scripts to change the current setting. It instead injects new content scripts in all tabs passing the setting as a content script option, based on which the DOM element is inserted/changed. `privly.js` looks at the DOM element and then decides if it needs to run or not.

Comment (smcgregor) [inline shared/javascripts/context_messenger.js]: I think it would be better to use messaging for this in the future, but that is not a problem for this pull request.

File: shared/javascripts/options.js [hunk 4]
```diff
@@ -63,6 +63,16 @@ if (Privly === undefined) {
   }
   Privly.options = {};
 
+  // CommonJS Module
+  if (typeof module !== "undefined" && module.exports) {
+    module.exports.options = Privly.options;
+    // load dependencies
+    ls = require("./local_storage.js").ls;
+    Privly.storage = require("./storage.js").storage;
```
Comment (smcgregor) [inline shared/javascripts/options.js]: Won't these already by on the `Privly` object when their scripts are included in the HTML page?

Comment (hitesh96db) [inline shared/javascripts/options.js]: This is when the script will be loaded as a CommonJS module in the background script. The background script will include a `require("chrome/content/privly-applications/shared/javascripts/options.js")` statement. There isn't any HTML page here.

Comment (smcgregor) [inline shared/javascripts/options.js]: So the options script coordinates the loading of all the modules? It may be better to include a separate script that imports the other scripts.

Comment (hitesh96db) [inline shared/javascripts/options.js]: This script is loaded [here](https://github.com/hitesh96db/privly-jetpack/blob/options_interface/lib/privly_ns.js#L8) as a CommonJS module.
Every script needs to load it's own dependencies. Won't there be a cross-dependency(as in circular reference) issue ?

Comment (hitesh96db) [inline shared/javascripts/options.js]: Also, while loading the module, all the dependencies have to be resolved.

File: shared/javascripts/context_messenger.js (hunk 3 above)
Comment (hitesh96db) [inline shared/javascripts/context_messenger.js]: I haven't tried it but that should cause the problem you were talking about(i.e, switching out of a tab and deactivating the injection). We have to remove the workers on "pagehide" and "detach" or else it'll cause the same problem you mentioned about [here](https://github.com/privly/privly-jetpack/pull/6#issuecomment-122185645)

Comment (smcgregor) [inline shared/javascripts/context_messenger.js]: OK, I'll classify this as "something to remember for the future"

File: shared/javascripts/options.js (hunk 4 above)
Comment (smcgregor) [inline shared/javascripts/options.js]: OK, and the `require` statement won't clear the state of the object when the object is reassigned in that manner?

Comment (hitesh96db) [inline shared/javascripts/options.js]: State as in ? It's similar to loading a script before another script in an HTML page.

Comment (smcgregor) [inline shared/javascripts/options.js]: `ls = require("./local_storage.js").ls;` assigns to the global variable `ls`. Assuming the `ls` variable has some property assigned to it already, like `ls.hello = "world";`, then the `require` statement doesn't clobber that property?

Comment (hitesh96db) [inline shared/javascripts/options.js]: Oh, the property won't be defined.

Comment (hitesh96db) [inline shared/javascripts/options.js]: But there isn't any problem here with that behavior.

Comment (smcgregor) [inline shared/javascripts/options.js]: So under this approach if there are `n` script files that all reference each other, there will be `n^2` script inclusions?

Comment (hitesh96db) [inline shared/javascripts/options.js]: n + (n-1) + (n-2) + (n-3) + ..... + 1 because circular reference is not possible.

Comment (smcgregor) [inline shared/javascripts/options.js]: OK, `n!`. Still, this seems odd from a performance perspective. Perhaps it would be good to check whether the object is already defined, then `require` it when needed. These variables are in global namespace so they can be shared.

Comment (hitesh96db) [inline shared/javascripts/options.js]: Well, there won't be anything already defined for a module. A CommonJS module can be thought of a black box, sort of separated from the outside env, or already loaded scripts. Whatever the black box needs, it asks for, and resolves the dependencies.

But there's still a certain level of optimization that can be done. The current way to define the Privly namespace in the background script is  this [file](https://github.com/hitesh96db/privly-jetpack/blob/options_interface/lib/privly_ns.js). But we know that `glyph.js` loads `options.js`, and `options.js` loads `storage.js` and `context_messenger.js`. So, we can make `options.js` export the loaded `storage.js` and `context_messenger.js`, all three of which now becomes available to `glyph.js` and then `glyph.js` can export all three and itself to the background script. So , [this](https://github.com/hitesh96db/privly-jetpack/blob/options_interface/lib/privly_ns.js) file which requires each of them separately can now require all of them using only `glyph.js`. 
I hope that made some sense ?

So we can optimize it at the background script but I doubt we can do anything about the individual modules. It's still a step forward if this works out :)

**Comment by hitesh96db** [comment]
@smcgregor Pushed a new commit.

**Comment by hitesh96db** [comment]
@smcgregor Is this Ok ?

**Comment by hitesh96db** [comment]
Thanks :)
